### DISCUSSION TITLE
$transaction, $executeRaw, and PrismaPromise Type Woes

### DISCUSSION THREAD

**Discussion by jeffdc** [prisma/prisma#7284] [Q&A] [answered]
I am in a TypeScript environment using Prisma 2.2.0. I am seriously frustrated with the `PrismaPromise` type and the way the API around `$executeRaw` and `$transaction` work. Hopefully I am just missing something obvious here.

To recap my understanding to make sure it is correct:

* calling `$transaction` requires an array of `PrismaPromise`
* calling `deleteMany` returns a `Prisma.BatchPayload` which is just a wrapper around a number
* calling `$executeRaw` returns a number

My code looks like this:
```typescript
const hostDeleteSteps = (speciesids: number[]): PrismaPromise<Prisma.BatchPayload>[] => {
    return [
        prisma.host.deleteMany({
            where: { host_species_id: { in: speciesids } },
        }),
        // ... several more deleteMany calls elided for clarity

        // I am trying to add this call and it is causing the type headaches
        prisma.$executeRaw(`DELETE FROM ... details elided for clarity;`,)
            .then((r) => {
                ({ count: r });
            }) as PrismaPromise<Prisma.BatchPayload>,
    ];
};
```

This is called by the following function. This code uses [fp-ts](https://gcanti.github.io/fp-ts/), the only thing that matters is that `TaskEither` wraps a `Promise` that might fail and `pipe` passes the result of executing each of its argument functions as an argument to the next function. So, the code executes the `deleteImages` function, then if that succeeds, executes the Prisma transaction as defined by `deleteHostTx` and then maps that result to a custom `DeleteResult` type that sums up the count of total records deleted.

```typescript
export const deleteHost = (speciesid: number): TaskEither<Error, DeleteResult> => {
    const deleteImages = () => TE.tryCatch(() => deleteImagesBySpeciesId(speciesid), handleError);
    const deleteHostTx = () => TE.tryCatch(() => db.$transaction(hostDeleteSteps([speciesid])), handleError);

    const toDeleteResult = (batch: Prisma.BatchPayload[]): DeleteResult => {
        return {
            type: 'host',
            name: 'host',
            count: batch.reduce((acc, v) => acc + v.count, 0),
        };
    };

    return pipe(
        deleteImages(),
        TE.chain(deleteHostTx),
        TE.map(toDeleteResult)
    );
};
```

My problem is that the calling `then` on a `PrismaPromise` changes the type to `Promise`. `PrismaPromise` is the only type that `$transaction` will accept. I can not kludge a fix because the Prisma client defines `PrismaPromise` with an [internal `unique` const](https://github.com/prisma/prisma/blob/ef2bfe4ea8fdbafe867cd1ac0bbbaced0d9851a3/src/packages/client/src/generation/TSClient/common.ts#L101) that I can not (sanely) replicate.

What is the recommended way to handle this? It really surprises me that a custom `PrismaPromise` type is needed.

**Comment by Rykuno** [comment]
Issue same on 2.29.1 as well. Currently experiencing this.

**Comment by yanavery** [comment]
@jeffdc, I know this is an old thread, but how did you overcome this issue? Running into something similar ... Thanks for sharing!

**Reply by jeffdc** [reply, marked as answer]
I kludged in a fix that circumvented the issue in a way that is highly specific to my use case.
https://github.com/jeffdc/gallformers/blob/9a696fcb3b61c08ff832ae9ab5ea9b76f28135a0/libs/db/host.ts#L405

**Comment by yanavery** [comment]
Ok, will have a look at it. Thanks for sharing!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
